hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSink.java

```diff
@@ -178,16 +179,19 @@ private void decorateConf() {
   /**
    * Replicate this array of entries directly into the local cluster using the native client. Only
    * operates against raw protobuf type saving on a conversion from pb to pojo.
+   * @param entries                    WAL entries to be replicated.
+   * @param cells                      cell scanner for iteration.
    * @param replicationClusterId       Id which will uniquely identify source cluster FS client
    *                                   configurations in the replication configuration directory
    * @param sourceBaseNamespaceDirPath Path that point to the source cluster base namespace
    *                                   directory
    * @param sourceHFileArchiveDirPath  Path that point to the source cluster hfile archive directory
+   * @param rsServerHost               regionserver coproc host.
    * @throws IOException If failed to replicate the data
    */
   public void replicateEntries(List<WALEntry> entries, final CellScanner cells,
     String replicationClusterId, String sourceBaseNamespaceDirPath,
-    String sourceHFileArchiveDirPath) throws IOException {
+    String sourceHFileArchiveDirPath, RegionServerCoprocessorHost rsServerHost) throws IOException {
```

Comment:
I suppose the RegionServerCroprocessorHost should be initialized when creating ReplicationSink?

I checked the code, maybe in ReplicationSinkServiceImpl, we could check whether the Server is an instance of HRegionServer, if so, we get its RegionServerCoprocessorHost for creating the ReplicationSink. And we could also reuse the AsyncClusterConnection of the Server, instead of creating a new one here.

Reply:
Correct, that's what I have added as part of this patch:

```
  @Override
  public void replicateLogEntries(List<AdminProtos.WALEntry> entries, CellScanner cells,
    String replicationClusterId, String sourceBaseNamespaceDirPath,
    String sourceHFileArchiveDirPath) throws IOException {
    RegionServerCoprocessorHost rsServerHost = null;
    if (server instanceof HRegionServer) {
      rsServerHost = ((HRegionServer) server).getRegionServerCoprocessorHost();
    }
    this.replicationSink.replicateEntries(entries, cells, replicationClusterId,
      sourceBaseNamespaceDirPath, sourceHFileArchiveDirPath, rsServerHost);
  }

```

So we do check for server instance to be regionserver, only then we pass non-null rs coproc host instance to this method.

Reply:
I mean we should pass RegionServerCoprocessorHost in when creating ReplicationSink, and store it as the a class field, so we do not need to pass it everytime when calling replicateEntries...

Reply:
Ah I see, yes that sounds good, will make these changes.

Reply:
This is addressed. Thanks @Apache9!